### PR TITLE
Combine some related compile functions together to save code size

### PR DESCRIPTION
**Pull request by dpgeorge**
A simple space saving optimisation that reduces the number of compile dispatch functions by combining common ones together.

Change in code size is:
```
   bare-arm:   -92 
minimal x86:  -544 
   unix x64:  -544 
unix nanbox:  -712 
      stm32:  -116 
     cc3200:  -128 
    esp8266:  -348 
      esp32:  -232
```

### CODE REVIEW

File: py/compile.c [hunk 1]
```diff
@@ -2102,16 +2093,16 @@ STATIC void compile_star_expr(compiler_t *comp, mp_parse_node_struct_t *pns) {
...
-
-STATIC void compile_and_expr(compiler_t *comp, mp_parse_node_struct_t *pns) {
-    c_binary_op(comp, pns, MP_BINARY_OP_AND);
+STATIC void compile_binary_op(compiler_t *comp, mp_parse_node_struct_t *pns) {
+    MP_STATIC_ASSERT(MP_BINARY_OP_OR + PN_xor_expr - PN_expr == MP_BINARY_OP_XOR);
+    MP_STATIC_ASSERT(MP_BINARY_OP_OR + PN_and_expr - PN_expr == MP_BINARY_OP_AND);
+    mp_binary_op_t binary_op = MP_BINARY_OP_OR + MP_PARSE_NODE_STRUCT_KIND(pns) - PN_expr;
+    int num_nodes = MP_PARSE_NODE_STRUCT_NUM_NODES(pns);
+    compile_node(comp, pns->nodes[0]);
+    for (int i = 1; i < num_nodes; i += 1) {
```
Comment (stinos) [inline py/compile.c]: Not sure if it's worth changing, but the rest of the code base almost exclusively uses `++i` or `i++`

**Comment by stinos** [comment]
I thought this size decrease would come at the cost of the code being less clear but that's not really the case. So this is a nice way to save space.

File: py/compile.c (hunk 1 above)
Comment (dpgeorge) [reply py/compile.c]: Good idea, now done.

**Comment by dpgeorge** [comment]
Thanks for the feedback.  I only merged those functions that had some common code, otherwise, you're right, it would have gotten messy.

**Comment by dpgeorge** [comment]
Ok, this was merged, see last commit 25ae98f07cb3c4488cb955403dfe56b8bb8db6f0

**Pull request closed without merging**